### PR TITLE
add package lime-paste fix #423

### LINKED ISSUE #423: Proposal: Simple reporting command for helping debug
Just an idea: it could be useful to have a command for helping bug reporting.
This would get debugging info from various sources and output it in an agglomerated form, optionally to a file.
It could be included in the `lime-debug` packages.

A list of data that could be included:

- /etc/config/lime (removing passwords)
- /etc/config/lime-defaults
- /etc/config/network
- /etc/config/wireless (removing passwords)
- dmesg
- batctl if
- batctl o
- bmx6 -c show=status show=interfaces show=links show=originators show=tunnels
- free
- ps
- ip address show
- ip route show
- brctl show
- ip link show
- /proc/cpuinfo
- df
- logread -l 20 -f
- iw dev wlan0-mesh station dump
- iw dev wlan1-mesh station dump
- iw dev wlan0-mesh mpath dump
- iw dev wlan1-mesh mpath dump

Additional interesting data can be:

- LibreMesh and OpenWrt version
- BATMAN adv logs, maybe it can be to retrieve with `batctl loglevel all; batctl log; batctl loglevel none`?
- ...?

### PR DESCRIPTION
**Pull request by aparcar**

### CODE REVIEW

**Comment by ilario** [comment]
Looks great :D (I still have to try it but I trust you did)

You could also have used the [spunge package](https://github.com/libremesh/lime-packages/tree/master/packages/sprunge) for doing the paste action and using this new package just for the report collecting. What do you think about this split of functions?

Anyway sprunge.us is currently offline (I'm trying to understand if this is permanent, if so we will need to modify, or even rename, the sprunge package to use the one you suggested, termbin.com, which looks good).

Could you expand this package's description to clarify the difference between this complete debugging package from the paste utility from sprunge package?

**Comment by aparcar** [comment]
@ilario I'd like the idea to have our own service running and not depend on sprunge.us, however if it's preferred I can ad a dependency to the sprunge package and use the service.

> Could you expand this package's description to clarify the difference between this complete debugging package from the paste utility from sprunge package?

Could you please provide such a description and I'll merge it?

**Comment by aparcar** [comment]
Runs now on chef.libremesh.org/paste

**Comment by ilario** [comment]
> I'd like the idea to have our own service running and not depend on sprunge.us,

Yess, that's ok, I agree to rely on whatever service is more reliable, and ours could well be.

What I was proposing is to separate the report generation code and the online submission routine. Just for modularity.

Then the submission routine could be based on our sprunge package (it already [uses netcat for submission](https://github.com/libremesh/lime-packages/blob/master/packages/sprunge/src/sprunge.sh#L45), it is very straight forward to adapt to use our server) renaming it to lime-pastebin.

The report generation routine can be based on this PR and call to lime-pastebin, when the -p option is used, for online submission. I would rename it to something like lime-report.

> Runs now on chef.libremesh.org/paste

It is amazing to have it on chef.libremesh.org but could we use another port?
I mean, for me right now, using netcat towards 9999 fails silently due to the shitty network configuration I'm on (filtered ports). Shitty networks are quite common, for example [eduroam also filters this port](https://www.eduroam.org/wp-content/uploads/2016/05/GN3-12-192_eduroam-policy-service-definition_ver28_26072012.pdf).
The ports that are (quite) never filtered and over 1k are: 3128, 8080, 1723 and 10000. Can we use 8080 or 3128 for the submission?

**Comment by aparcar** [comment]
let's wait for a comment of @G10h4ck

**Comment by p4u** [comment]
Looks good. The only comment I can do is to use another subdomain, for instance "paste.libremesh.org". It can just be for now a CNAME to chef.libremesh.org but in the future we might want to separated both services.

**Comment by G10h4ck** [comment]
Seems nice!

I would rather use lime-paste as a generic command with functionalities as sprunge but based on our infrastructure, and then rename this one to lime-report that paste this lime related files and commands

Cheers!

**Comment by aparcar** [comment]
fair, @p4u please setup the domain. I'll split the packages

**Comment by p4u** [comment]
I've created paste.libremesh.org and report.libremesh.org

**Comment by ilario** [comment]
Looks good :)
Some comment:

- as I mentioned above (and seems in agreement with Pau) I would use a port that is never filtered (like 8080, or 80 if it is easy to do), but it is not a blocker, 9999 is also fine;
- in both the Makefile you used INSTALL_DATA instead of INSTALL_BIN, the second one is more appropriate as it sets the +x and both lime-paste and lime-report are intended to be executable;
- `lime-report` prints the help and this is ok, the problem is that `lime-report -h` does not print the help;
- in lime-report, the variable $PASTE_SERVICE is not set, I would replace it with "paste.libremesh.org" string;
- in the logread command, the `-f` option "Follow log messages" should not be used, otherwise the script hangs at that point (it's my fault, I included it in #423);
- in lime-report command, the `-f` option could print the filename to STDOUT, not strictly needed but could be nice;
- in busybux's echo command, the `echo "\n"` just prints a \n instead of a newline. So in lime-report the `echo "\nSOMETHING\n"` commands could be replaced with `echo -e "\nSOMETHING\n"` or with `echo; echo "SOMETHING"; echo`;
- what about including also output of `iwinfo`? iwinfo includes also interfaces which are not explicitly listed in the iw commands;
- I have not a clue about why but trying the `lime-report -p` makes available online just the first 32767 characters of the report, I fear it is a server issue as using termbin.com as PASTE_SERVICE shows the complete paste. Compare [this on our server](http://paste.libremesh.org/xxyu/) with [this on termbin.com](https://termbin.com/3lcs);
- what I suggested (to adapt the sprunge package instead of creating lime-paste) was actually a bas idea as, even if netcat is used the submission method is different to the [Fiche](https://github.com/solusipse/fiche) one, the current solution in lime-paste is great. For the evolution of sprunge package see issue #458.

**Comment by ilario** [comment]
>>    as I mentioned above (and seems in agreement with Pau) I would use a port that is never filtered (like 8080, or 80 if it is easy to do), but it is not a blocker, 9999 is also fine;
> I hope @p4u implements that, not sure how this actually works with nginx

I suggest to just use the -p option from fiche, much easier :) (with that option one can set 8080 but cannot set 80 unless running with root)

>>    in lime-report, the variable $PASTE_SERVICE is not set, I would replace it with "paste.libremesh.org" string;
> why and where? The variable belongs to lime-paste

It is used in lime-report help, in the line:
```
     echo "-p    paste to $PASTE_SERVICE" && \
```

>>    I have not a clue about why but trying the lime-report -p makes available online just the first 32767 characters of the report, I fear it is a server issue as using termbin.com as PASTE_SERVICE shows the complete paste. Compare this on our server with this on termbin.com;
> I changed the buffer size to 16kb, hope that works

Increase it, don't decrease it :PPPP
It is set with the -B fiche option, right? On my router the lime-report is 48 kB, so I would set at least 64 kB.

> Let's remove the sprunge.us package and use lime-paste. We would have to rename it anyway, so why not just use lime-paste?

For sure sprunge or whatever name will not be included in lime-full anymore. Anyway let's keep the topic on #458

**Comment by aparcar** [comment]
> I suggest to just use the -p option from fiche, much easier :) (with that option one can set 8080 but cannot set 80 unless running with root)

I think @p4u plans to filter, if a http header comes in, forward to the paste folder, if not, forward to localhost:9999 aka fiche. Allowing to run both on port 80, sending and receiving. However I haven't done this before and therefore need his help.

>  It is set with the -B fiche option, right? On my router the lime-report is 48 kB, so I would set at least 64 kB.

whoops, yea now it's 64kB

**Comment by ilario** [comment]
One more thing:
can you add lime-paste and lime-report and remove sprunge in [lime-debug Makefile](https://github.com/libremesh/lime-packages/blob/master/packages/lime-debug/Makefile#L26)?

Then there's just the used but undefined $PASTE_SERVICE in lime-report and the port doubt and is perfect :)

Last thought: maybe we should truncate the dmesg output for avoiding the risk to fill fiche buffer with long uptime router's dmesg? Or we can make the fiche buffer very huge?

**Comment by aparcar** [comment]
@ilario all done, see here https://github.com/libremesh/lime-packages/compare/32884f150f51d6b050deaf0688b6877e1ef3f60d..1ab15a438d22c966717ec1055d8b1339493524e8

**Comment by ilario** [comment]
All amazing!
I didn't even know that fallback options like ${PASTE_PORT:-8080} were possible :D
Thanks for going through all these iterations <3

I have one last (last, I promise :) ) observation: `dmesg | tail -c 10` shows just the last 10 characters which is too little! In my opinion we can use `dmesg | tail -n 400` so that it cuts just very long and crazy dmesg (approximatively it cuts dmesg outputs over 30 kB, safe value for fiche buffer that now has been increased) but it does not cut a regular dmesg.

**Comment by aparcar** [comment]
oh we should change it to -c 30000

**Comment by ilario** [comment]
Yess, that also would be ok

**Comment by p4u** [comment]
I got some fun and end up with this https://github.com/p4u/spaste

It's already running, you can test it (use spaste.sh script).

`uci show lime | sh spaste.sh lime123`

See content on-line:

`http://paste.libremesh.org/get/lime123`

**Comment by p4u** [comment]
@aparcar would you like to adapt the package to use spaste.sh (rename it to lime-paste) ?

**Comment by ilario** [comment]
`laspalmeras` sounds like a safe enough password @p4u
The new spaste looks good but is there a good reason for preferring it over fiche?
This does not mean we cannot have them both, e.g. paste.libremesh.org with fiche and spaste.libremesh.org with spaste, on two different servers for ensuring availability.

**Comment by p4u** [comment]
This password is not secret :)

The biggest difference is that it uses standard HTTP and so can run behind an HTTP server proxy, so we can use port 80 or even 443 with SSL enabled (if curl/wget and OpenSSL available in the router).

I don't see the point of having both. I'd instead choose one. Let's see what @aparcar says.

**Comment by p4u** [comment]
More diferencies are:

+ spaste allows specify your own paste id name and so rewrite/update pastes
+ spaste uses levelDB key/value datastore, much faster than raw I/O disk

File: packages/lime-paste/files/lime-paste.sh [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+
+PASTE_SERVICE="${PASTE_SERVICE:-paste.libremesh.org}"
+PASTE_PORT="${PASTE_PORT:-8080}"
+
+nc "$PASTE_SERVICE" "$PASTE_PORT"
```
[the remark this answers is not in the document]
Comment (aparcar) [reply packages/lime-paste/files/lime-paste.sh]: I did not fail on my local machine...

**Comment by aparcar** [comment]
@p4u cool work, please adapt the lime-paste and report package. I really don't care so much what to use.

> spaste allows specify your own paste id name and so rewrite/update pastes

Problem could become that people all think "error" or "report" is a good ID and and randomly overwrite pastes of others. Also this is kind of a attack vector that you can change all links posted online to your desired content. Not?

> spaste uses levelDB key/value datastore, much faster than raw I/O disk

So we should store our repo in levelDB as well? ;)

**Comment by p4u** [comment]
> @p4u cool work, please adapt the lime-paste and report package. I really don't care so much what to use.
> 
> > spaste allows specify your own paste id name and so rewrite/update pastes
> 
> Problem could become that people all think "error" or "report" is a good ID and and randomly overwrite pastes of others. Also this is kind of a attack vector that you can change all links posted online to your desired content. Not?
> 
> > spaste uses levelDB key/value datastore, much faster than raw I/O disk
> 
> So we should store our repo in levelDB as well? ;)

It's not supposed to be a very secure nor private nor permanent service. For instance a much worse attack vector (which affects both solutions) is to fill all hard disk space by writing random stuff.

Anyway if paste name is not specified, spaste server uses a self calculated semi-random ID. So we might just not add the ID by default on lime-report.

File: packages/lime-paste/files/lime-paste.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (p4u) [reply packages/lime-paste/files/lime-paste.sh]: Sure it is, your local machine netcat has nothing to do with busybox nc.

**Comment by G10h4ck** [comment]
What is the status of this PR, I agree with @p4u that using HTTP as transport is important because this way the server can run on port 80 while sharing the public IP with other applications/websites. Also it would be really nice if lime-paste could be used without us maintaining yet another server no one want to maintain, without introducing additional dependencies into the router.

**Comment by G10h4ck** [comment]
The sprunge package can be used as an example on how to use busybox netcat to emulate an HTTP client

**Comment by p4u** [comment]
ping @aparcar

**Comment by p4u** [comment]
> The sprunge package can be used as an example on how to use busybox netcat to emulate an HTTP client

That's already solved, please see this https://github.com/p4u/spaste/blob/master/spaste.sh

**Comment by p4u** [comment]
Actually everything is prepared, even the HTTP service is already running, for instance http://paste.libremesh.org/get/hello

**Comment by aparcar** [comment]
Sorry please do whatever you like regarding pastes

**Comment by p4u** [comment]
> I'd prefer to reuse your PR, just ammend and force push. It's as easy as changing the contents of `lime-paste.sh` by `spaste.sh` (copy&paste).

Can you just amend and push?

**Comment by aparcar** [comment]
@p4u looks like I can't reopen this

**Comment by aparcar** [comment]
https://github.com/libremesh/lime-packages/pull/457.diff

**Comment by ilario** [comment]
I totally agree with @G10h4ck on:

> it would be really nice if lime-paste could be used without us maintaining yet another server no one want to maintain

But as even the existing services have quite a lot of downtimes (and ours could have the same problem), the solution I propose is to give the user multiple options, the user will use the one which works in that moment.

For example `lime-report -p1` could paste using spaste; `lime-report -p2` could paste using lime-paste as currently included in this PR; `lime-report -p3` could paste using ix.io service; `lime-report -p4` could paste using sprunge.us service.

**Comment by ilario** [comment]
If @aparcar is ok with this, I am going to open another PR based on @aparcar's code of lime-report.
I'll remove the paste feature from it which I think it should be managed externally to lime-report.
If we want to also resurrect lime-paste I can but I'll keep the two packages independent, let me know.

**Pull request closed without merging**